internal/fs/dir_handle.go

```diff
@@ -156,78 +174,79 @@ func fixConflictingNames(entries []fuseutil.Dirent) (err error) {
-	// like inode IDs based on (object name, generation) hashes. But then what
-	// about the birthday problem? And more importantly, what about our
-	// semantic of not minting a new inode ID when the generation changes due
-	// to a local action?
-	for i, _ := range entries {
-		entries[i].Inode = fuseops.RootInodeID + 1
-	}
-
-	return
-}
+// Fetch Dirent entries from GCSfuse.Will be used as a goroutine which is run asynchronously
+// to fetch data in the background while kernel requests are also served simultaneously.
+func (dh *DirHandle) FetchEntriesAsync(
+	rootInodeId int,
+	firstCall bool) {
+	ctx, cancel := context.WithCancel(context.Background())
+	var err error
+	//ContinuationToken is also empty in case of firstCall and after all entries have been fetched.
+	//Keep fetching entries in batches of MaxResultsForListObjectsCall
+	for ContinuationToken != "" || firstCall {
+		var entries []fuseutil.Dirent
+		dh.In.Lock()
+		entries, ContinuationToken, err = dh.In.ReadEntries(ctx, ContinuationToken)
+		dh.In.Unlock()
+		if err != nil {
+			err = fmt.Errorf("ReadEntries: %w", err)
+			rec.Lock()
+			rec.err = err
+			rec.Unlock()
+			//Signal the suspended go routine that an error has occurred.
+			rec.cond.Broadcast()
+			//cancel the context to release the associated resources
+			cancel()
+			//cancelling the context does not kill the go routine.Killing
+			//to prevent go routine leak.
+			runtime.Goexit()
+		}
+		//Use the last entry from the last fetch for fixing naming conflicts
+		if !firstCall {
+			rec.Lock()
+			entries = append(entries, rec.entryForSorting)
+			rec.Unlock()
+		}
+		sort.Sort(sortedDirents(entries))
+		err = fixConflictingNames(entries)
 
-// LOCKS_REQUIRED(dh.Mu)
-// LOCKS_EXCLUDED(dh.in)
-func (dh *dirHandle) ensureEntries(ctx context.Context) (err error) {
-	dh.in.Lock()
-	defer dh.in.Unlock()
-
-	// Read entries.
-	var entries []fuseutil.Dirent
-	entries, err = readAllEntries(ctx, dh.in)
-	if err != nil {
-		err = fmt.Errorf("readAllEntries: %w", err)
-		return
-	}
+		if err != nil {
+			err = fmt.Errorf("fixConflictingNames: %w", err)
+			rec.Lock()
+			rec.err = err
+			rec.Unlock()
+			//Signal the suspended go routine that an error has occurred.
+			rec.cond.Broadcast()
+			cancel()
+			runtime.Goexit()
+		}
 
-	// Update state.
-	// Fix up offset fields.
-	for i := 0; i < len(entries); i++ {
-		entries[i].Offset = fuseops.DirOffset(uint64(len(dh.entries) + i + 1))
+		//Save the last entry from current fetch to use it for
+		//fixing naming conflicts for next fetch
+		if ContinuationToken != "" {
+			rec.Lock()
+			rec.entryForSorting = entries[len(entries)-1]
+			rec.Unlock()
+			entries = entries[:len(entries)-1]
+		}
+		rec.Lock()
+		//Update InodeID and Offset for the entries
+		for i := range entries {
+			entries[i].Inode = fuseops.InodeID(rootInodeId + 1)
+			entries[i].Offset = fuseops.DirOffset(uint64(rec.length + i + 1))
+		}
+		rec.length += len(entries)
+		rec.Unlock()
+
+		dh.Mu.Lock()
```

Comment:
this looks wierd, you are storing same data in two variables under 2 different locks - dh.entries and rec.length. Will be difficult to understand and maintain.